src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,144 @@
+            last.next = addedElement;
+            last = addedElement;
+        }
+        size++;
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException();
+        }
+        if (index == size) {
+            add(value);
+            return;
+        }
+        if (index == 0) {
+            MyNode<T> current = new MyNode<>(null, value, first);
+            MyNode<T> nextFromCurrent = first;
+            nextFromCurrent.prev = current;
+            first = current;
+        } else {
+            MyNode<T> current = find(index);
+            MyNode<T> added = new MyNode<>(current.prev, value, current);
+            current.prev.next = added;
+            current.prev = added;
+        }
+        size++;
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException();
+        }
+        return find(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException();
+        }
+        MyNode<T> current = find(index);
+        T oldValue = current.item;
+        current.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index == size) {
+            throw new IndexOutOfBoundsException();
+        }
+        MyNode<T> removed = find(index);
+        T result = removed.item;
+        if (index == 0) {
+            if (size == 1) {
+                first = null;
+                first = last;
+            } else {
+                first = removed.next;
+                first.prev = null;
+            }
+        } else if (index == size - 1) {
+            last = removed.prev;
+            last.next = null;
+        } else {
+            removed.next.prev = removed.prev;
+            removed.prev.next = removed.next;
+        }
+        size--;
+        return result;
     }
 
     @Override
     public boolean remove(T object) {
+        MyNode<T> current = first;
+        int i = 0;
+        for (i = 0; i < size; i++) {
+            if (object == current.item || object != null && object.equals(current.item)) {
+                remove(i);
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private MyNode<T> find(int index) {
```

Comment:
let's iterate list from head to middle or from tail to middle depending on the index